src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -371,12 +371,16 @@ private ComponentState AttachAndInitComponent(IComponent component, int parentCo
         var parentComponentState = GetOptionalComponentState(parentComponentId);
         var componentState = CreateComponentState(componentId, component, parentComponentState);
         Log.InitializingComponent(_logger, componentState, parentComponentState);
-        _componentStateById.Add(componentId, componentState);
-        _componentStateByComponent.Add(component, componentState);
```

Comment:
Moved this into a method that `ComponentState` calls in its constructor to set this up, so that cascading value providers can access the RenderMode during the initial setup.